Adding wrap/unwrapCryptoKey to ServiceWorker

nmahendru opened the pull request:
#### 51ae1a6d501e745cb0af429b3323372ce9aaae62
<pre>
Adding wrap/unwrapCryptoKey to ServiceWorker
<a href="https://bugs.webkit.org/show_bug.cgi?id=183167">https://bugs.webkit.org/show_bug.cgi?id=183167</a>
<a href="https://rdar.apple.com/88339459">rdar://88339459</a>

Reviewed by Sihui Liu.

At the moment the ServiceWorker flow does not have the wrap and unwrap
functionality so any attempt to write a CryptoKey from ServiceWorker to indexDB
is failing.

This introduces the bridge from SW all the way to UIProcess so that it can call the
key wrapping APIs and serialize keys for IndexedDB.

* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https-expected.txt: Added.
* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https.html: Added.
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js: Added.
(async createKeys):
(async addValue):
(async getValue.):
(async getValue):
(async testFetchedKeys):
(async runTest):
(async cleanupDatabase.):
(async cleanupDatabase):
(async self):
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get.js: Added.
(async test):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::wrapCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::wrapSerializedCryptoKey):
(WebCore::unwrapSerializedCryptoKey):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::wrapCryptoKey):
(WebCore::std::optional&lt;Vector&lt;uint8_t&gt;&gt;Document::unwrapCryptoKey):
(WebCore::Document::unwrapCryptoKey): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::wrapCryptoKey const): Deleted.
(WebCore::ChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::webCryptoClient):
(WebCore::Page::webCryptoClient const):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/WebCryptoClient.cpp: Added.
* Source/WebCore/page/WebCryptoClient.h: Added.
(WebCore::WebCryptoClient::wrapCryptoKey const):
(WebCore::WebCryptoClient::unwrapCryptoKey const):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::wrapCryptoKey):
(WebCore::WorkerGlobalScope::unwrapCryptoKey):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::wrapCryptoKey):
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::wrapCryptoKey):
(WebKit::WebProcessProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::_webCryptoMasterKeyForWebProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/WebCoreSupport/SWWebCryptoClient.cpp: Added.
(WebKit::SWWebCryptoClient::wrapCryptoKey const):
(WebKit::SWWebCryptoClient::unwrapCryptoKey const):
* Source/WebKit/WebProcess/WebCoreSupport/SWWebCryptoClient.h: Added.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::wrapCryptoKey const): Deleted.
(WebKit::WebChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.cpp: Added.
(WebKit::WebPageWebCryptoClient::wrapCryptoKey const):
(WebKit::WebPageWebCryptoClient::unwrapCryptoKey const):
(WebKit::WebPageWebCryptoClient::WebPageWebCryptoClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.h: Added.
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
* Source/WebKitLegacy/WebCoreSupport/WebViewWebCryptoClient.h: Added.
* Source/WebKitLegacy/WebCoreSupport/WebViewWebCryptoClient.mm: Added.
(WebViewWebCryptoClient::wrapCryptoKey const):
(WebViewWebCryptoClient::unwrapCryptoKey const):
(WebViewWebCryptoClient::WebViewWebCryptoClient):
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::wrapCryptoKey const): Deleted.
(WebChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
(TEST):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate webCryptoMasterKeyForWebProcess]):

Canonical link: <a href="https://commits.webkit.org/276088@main">https://commits.webkit.org/276088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbb46773e7695cd38485adabfac0af0598fada50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46159 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46364 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39824 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46024 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20176 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36104 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44294 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37656 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17086 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17322 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38724 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1777 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39877 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47919 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15330 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42907 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41592 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9723 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/ServiceWorkerChromeClient.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace WebKit {
+
+class ServiceWorkerChromeClient:  public WebCore::EmptyChromeClient {
```
Conceptually `EmptyChromeClient` means a dumb client that does nothing (i.e. no client), so it seems not right to subclass it and provide partial functionalities.
Let's add a new class `CryptoKeyClient` (or some other names since you have better context) for the crypto key operations and remove them from ChromeClient. You could add `CryptoKeyClient` to the `PageConfiguration` and subclass it to provide implementation in WebKit (similar to what we did for `BroadcastChannelRegistry`)

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
```diff
@@ -72,4 +72,5 @@ WK_API_AVAILABLE(macos(10.15), ios(13.0))
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore domain:(NSString *)registrableDomain didOpenDomainViaWindowOpen:(NSString *)openedRegistrableDomain withProperty:(WKWindowProxyProperty)property directly:(BOOL)directly;
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore didAllowPrivateTokenUsageByThirdPartyForTesting:(BOOL)wasAllowed forResourceURL:(NSURL *)resourceURL;
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore domain:(NSString *)registrableDomain didExceedMemoryFootprintThreshold:(size_t)footprint withPageCount:(NSUInteger)pageCount processLifetime:(NSTimeInterval)processLifetime inForeground:(BOOL)inForeground;
+- (NSData *)_webCryptoMasterKeyForWebProcess;
```
Is the crypto master key going to change in run time? If not, I think `WKWebsiteDataStoreConfiguration` is a better place for this value, so we can ask client to provide it at the creation time instead of pulling it.

achristensen07 replied:
We currently only need to access the master key when a page uses crypto APIs.  Making this change would make it so that we always need to access the key from the keychain before we can show any page.

szewai commented on Source/WebCore/page/WebCryptoClient.h:
```diff
@@ -0,0 +1,42 @@
+class WebCryptoClient {
+    WTF_MAKE_FAST_ALLOCATED;
+protected:
+    WEBCORE_EXPORT WebCryptoClient();
```
Usually the "Web" prefix is used for WebKit classes, how about renaming this to `CryptoClient` or `CoreCryptoClient` and renaming the WebKit/WebKitLegacy classes to `WebCryptoClient`?

szewai commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2040,10 +2047,12 @@ class CloneSerializer : public CloneBase {
-                if (!wrapCryptoKey(m_lexicalGlobalObject, serializedKey, wrappedKey))
+
+                auto wrappedKey = wrapCryptoKey(m_lexicalGlobalObject, serializedKey);
+                if (!wrappedKey.has_value())
```
`std::optional` can be implicitly converted to bool here, so you could `if (!wrappedKey)`

szewai commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -5158,15 +5167,16 @@ class CloneDeserializer : public CloneBase {
-            Vector<uint8_t> serializedKey;
-            if (!unwrapCryptoKey(m_lexicalGlobalObject, wrappedKey, serializedKey)) {
+            auto serializedKey = unwrapCryptoKey(m_lexicalGlobalObject, wrappedKey);
+            if (!serializedKey.has_value()) {
```
`if (!serializedKey)`

szewai commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2040,10 +2047,12 @@ class CloneSerializer : public CloneBase {
                     return false;
-                write(wrappedKey);
+
+                write(wrappedKey.value());
```
`write(*wrappedKey)`

szewai commented on Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
```diff
@@ -211,6 +211,8 @@ bool deleteDefaultWebCryptoMasterKey()
 
 bool wrapSerializedCryptoKey(const Vector<uint8_t>& masterKey, const Vector<uint8_t>& key, Vector<uint8_t>& result)
 {
+    if (masterKey.isEmpty())
```
Does the code forbid using an empty master key before? If not, I wonder whether this could cause breakage on existing data.
For this function itself, it seems it does not prevent using an empty key.

nmahendru replied:
This check is failfast. If the masterKey is zero length there is no encryption to be done. So I have added that. The underlying crypto library will return the same error.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/SWWebCryptoClient.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace WebKit {
+
+class SWWebCryptoClient:  public WebCore::WebCryptoClient {
```
You don't need to add a new class for service worker, as it has the same behavior as `WebPageWebCryptoClient` -- sending message to UI process. What you could do is making `WebPageWebCryptoClient` (or renamed to `WebKit::WebCryptoClient`) store std::optional<WebCore::PageIdentifier>, to indicate whether there is a page, and send different messages based on whether page identifier is null.
If client provides the same master key for all pages, then we should remove the need for PageIdentifier too; but that could be done later.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.cpp:
```diff
@@ -0,0 +1,57 @@
+    if (!sendResult.succeeded())
+        return std::nullopt;
+
+    return std::get<0>(sendResult.takeReply());
```
```
auto [wrappedKey] = sendResult.takeReplyOr(std::nullopt);
return wrappedKey;
```

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.cpp:
```diff
@@ -0,0 +1,57 @@
+    if (!sendResult.succeeded())
+        return std::nullopt;
+
+    return std::get<0>(sendResult.takeReply());
```
```
auto [unwrappedKey] = sendResult.takeReplyOr(std::nullopt);
return unwrappedKey;
```

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.cpp:
```diff
@@ -0,0 +1,57 @@
+    return std::get<0>(sendResult.takeReply());
+}
+
+WebPageWebCryptoClient::WebPageWebCryptoClient(WebPage& page)
```
This could take std::optional<WebCore::PageIdentifier> directly

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,153 @@
+}
+
+self.addEventListener("activate", async () => {
+    await cleanupDatabase();
```
Actually I think WebKitTestRunner cleans up data between test runs so this might not be needed.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
```diff
@@ -72,4 +72,5 @@ WK_API_AVAILABLE(macos(10.15), ios(13.0))
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore domain:(NSString *)registrableDomain didOpenDomainViaWindowOpen:(NSString *)openedRegistrableDomain withProperty:(WKWindowProxyProperty)property directly:(BOOL)directly;
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore didAllowPrivateTokenUsageByThirdPartyForTesting:(BOOL)wasAllowed forResourceURL:(NSURL *)resourceURL;
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore domain:(NSString *)registrableDomain didExceedMemoryFootprintThreshold:(size_t)footprint withPageCount:(NSUInteger)pageCount processLifetime:(NSTimeInterval)processLifetime inForeground:(BOOL)inForeground wasPrivateRelayed:(BOOL)wasPrivateRelayed;
+- (NSData *)webCryptoMasterKeyForWebProcess;
```
Let's call it `webCryptoMasterKey` as I think the web process part is an implementation detail and we are not including process id in the method.

szewai commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -26,6 +26,7 @@
 import CryptoKit
 import Foundation
 
+
```
Unneeded?

szewai commented on Source/WebCore/page/CryptoClient.cpp:
```diff
@@ -0,0 +1,33 @@
+
+namespace WebCore {
+CryptoClient::CryptoClient() = default;
+CryptoClient::~CryptoClient() = default;
```
For one-liners we don't need to put them in cpp file. Let's move this to header and drop WEBCORE_EXPORT.

szewai commented on Source/WebCore/loader/EmptyClients.h:
```diff
@@ -237,4 +238,11 @@ class EmptyChromeClient : public ChromeClient {
 DiagnosticLoggingClient& emptyDiagnosticLoggingClient();
 WEBCORE_EXPORT PageConfiguration pageConfigurationWithEmptyClients(std::optional<PageIdentifier>, PAL::SessionID);
 
+class EmptyWebCryptoClient: public CryptoClient {
```
`EmptyCryptoClient`

szewai commented on Source/WebCore/platform/PageIdentifier.h:
```diff
@@ -21,6 +21,7 @@
  * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
  * THE POSSIBILITY OF SUCH DAMAGE.
+ 
```
Unneeded.

szewai commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -73,6 +73,7 @@
 #include "WorkerThread.h"
 #include <JavaScriptCore/ScriptArguments.h>
 #include <JavaScriptCore/ScriptCallStack.h>
+#include <cstdint>
```
Why is this needed?

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -83,14 +84,29 @@
         , m_hasWindowProxyPropertyAccessSelector([m_delegate.get() respondsToSelector:@selector(websiteDataStore:domain:didOpenDomainViaWindowOpen:withProperty:directly:)])
         , m_hasDidAllowPrivateTokenUsageByThirdPartyForTestingSelector([m_delegate.get() respondsToSelector:@selector(websiteDataStore:didAllowPrivateTokenUsageByThirdPartyForTesting:forResourceURL:)])
         , m_hasDidExceedMemoryFootprintThresholdSelector([m_delegate.get() respondsToSelector:@selector(websiteDataStore:domain:didExceedMemoryFootprintThreshold:withPageCount:processLifetime:inForeground:wasPrivateRelayed:)])
+        , m_hasWebCryptoMasterKey([m_delegate.get() respondsToSelector:@selector(webCryptoMasterKey)])
```
`m_hasWebCryptoMasterKeySelector`

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -83,14 +84,29 @@
     {
         if (!m_hasRequestStorageSpaceSelector || !m_delegate) {
-            completionHandler({ });
+
```
Why this change?

nmahendru replied:
oops.. reverting.. unintentional!

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -24,6 +24,7 @@
  */
 
 #import "config.h"
+#include <Foundation/Foundation.h>
```
This should be `import` and should be moved to below.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1933,6 +1938,107 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+@implementation SWCryptoKeyMessageHandler
+- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
+{
+    WTFLogAlways("%s\n", [[message body] UTF8String]);
```
We normally don't use `WTFLogAlways` unless for debugging. Please use EXPECT_* for checking the content.

nmahendru replied:
This is just a debugging aid I have left inside the test to help see the progress of the test for timeouts(in future?) or if it fails, this will print the javascript exception to the logs which is very useful to debug the tests from EWS logs.

szewai replied:
For debugging, I mean actively debugging an existing issue (e.g. the test are already flaky on bots and cannot be reproduced locally), and that should be a temporary measure (i.e. we will remove the log as soon as we figure out what's wrong, otherwise the bot's result will be noisy, and bot can be slowed down). This is not the case.

nmahendru replied:
> bot's result will be noisy, and bot can be slowed down

Thanks for explaining. I get that. I will remove it.

szewai commented on Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
```diff
@@ -48,4 +48,5 @@
 - (NSString*)lastUpdatedBackgroundFetchIdentifier;
 - (NSArray*)reportedWindowProxyAccessDomains;
 - (void)clearReportedWindowProxyAccessDomains;
+- (NSData *)webCryptoMasterKeyForWebProcess;
```
The name is not correct, and you don't need definition for this. You could just implement `webCryptoMasterKey` in .mm as other delegate methods. 
I don't think we need to implement `webCryptoMasterKey ` for TestWebsiteDataStoreDelegate, as it's not used in any test (the new test defines a new `CryptoKeyWebsiteDataStoreDelegate `).

nmahendru replied:
This is used in layout test that I have added.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1933,6 +1938,107 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+}
+@end
+
+static std::string readFile(std::string filename, std::string type)
```
We have FileSystem::readEntireFile in wtf/FileSystem.h.

nmahendru replied:
thanks! done.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1933,6 +1938,107 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+    done = false;
+
+    // assert check masterKey fetch
+    ASSERT_EQ(15ul, [dataStoreDelegate fetchCount]);
```
It seems this test mainly ensures the delegate method is called and the rest is the same as the layout test, so I don't think we need this API test and the layout test is enough.
If you want to test the delegate method, we should probably implement it WebKitTestRunner to make the test not use the default master key.

nmahendru replied:
I need an API test to make sure of two things in order to ensure that this is not causing a regression:
1. Make sure that this change does not crash the process until we add the delegate implementation in Safari.
2. Install the delegate and be able to assert that a C++ delegate API is called.

szewai replied:
Why do we need to specifically check there is no crash after this change? Does process crash before this change? 
For the second reason, you mentioned you already use the delegate method in WebKitTestRunner above, so you should be able to verify with the encrypted content in layout test.

szewai commented on Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
```diff
@@ -151,4 +151,12 @@ - (void)clearReportedWindowProxyAccessDomains
+    // Not so random key
+    constexpr size_t keyLength = 16;
+    uint8_t keyBytes[keyLength] = { 1 };
+    NSData * key = [NSData dataWithBytes:keyBytes length:(keyLength)];
```
`NSData *key`

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+async function addValue(key, value) {
+    return new Promise(
+        function (resolve, reject) {
+            var request = self.indexedDB.open(dbName, 1);
```
The test does not need to open one database connection for each request (which may slow down the test since it would require multiple async messages, ); it can open the database once and store it in global variable, something like (key path is optional):
```
async function openDatabase() 
{
    return new Promise((resolve, reject) => { 
        var request = self.indexedDB.open(dbName);
        request.onupgradeneeded = (event) => { event.target.result.createObjectStore(objstoreName); }
        request.onsuccess = (event) => { 
            resolve(event.target.result);
        }
        request.onerror = () => {
            reject(event.target.error.name);
        }
    });
}
```
and in `runTest()`:
```
db = await openDatabase();
```
This will also reduce nested statements in other functions.

nmahendru replied:
excellent suggestion! done.

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+                case "AES-GCM":
+                case "AES-CBC":
+                    var ciphertext = new Uint8Array(await self.crypto.subtle.encrypt({name: key, iv: iv}, value, data));
+                    if (!(ciphertext.length > 0)) {
```
Do you mean `if (!ciphertext.length)`?

nmahendru replied:
yes. changed.

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+                    break;
+                case "ECDSA":
+                    var signature = new Uint8Array(await self.crypto.subtle.sign({name: key, hash: "SHA-384"}, value.privateKey, data));
+                    if (!(signature.length > 0)) {
```
`if (!signature.length)`?

nmahendru replied:
done.

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+            await addValue(key, value);
+            fetchedKeys[key] = await getValue(key);
+        }
+        let check = await testFetchedKeys(fetchedKeys);
```
It seems we could move `testFetchedKeys` inside the loop and return early if some key fails the test, e.g.:
```
fetchedValue = await getValue(key);
result = await testFetchedValue(key, fetchedValue);
if (result.length)
    return new Response(null, {status: 200, statusText: "Test Failed: testFetchedValue fails on key - " + key});
```
With this change, `testFetchedValue` should take two parameters.

nmahendru replied:
done! failfast.

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+                    break;
+                case "HMAC":
+                    signature = new Uint8Array(await self.crypto.subtle.sign(key, value, data));
+                    if (!(signature.length > 0)) {
```
`if (!signature.length)`?

nmahendru replied:
done.

szewai commented on LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
```diff
@@ -0,0 +1,135 @@
+
+}
+
+async function testFetchedKeys(fetchedKeys) {
```
Let's call it `testFetchedValue(key, value)` and only check one value at a time. Then we could use it in the loop of `for (let [key, value] of Object.entries(keys))` in `runTest`.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -83,10 +84,25 @@
+    Vector<uint8_t> webCryptoMasterKey() final
+    {
+        if (!m_hasWebCryptoMasterKeySelector || !m_delegate)
+            return { };
```
It seems for `WebKitLegacy` and `NavigationState::NavigationClient::webCryptoMasterKey` we fall back to `defaultWebCryptoMasterKey()` when delegate method is not implemented. Is there a reason that we don't want to do that here?

nmahendru replied:
I think you are right. 
I have added the `defaultWebCryptoMasterKey` path here  as a fallback if the delegate is not set.
the same is also true for WebPageProxy::wrapCryptoKey implementation.

szewai reviewed (approved):
r=me with comments

nmahendru committed:
Adding wrap/unwrapCryptoKey to ServiceWorker
https://bugs.webkit.org/show_bug.cgi?id=183167
rdar://88339459

Reviewed by Sihui Liu.

At the moment the ServiceWorker flow does not have the wrap and unwrap
functionality so any attempt to write a CryptoKey from ServiceWorker to indexDB
is failing.

This introduces the bridge from SW all the way to UIProcess so that it can call the
key wrapping APIs and serialize keys for IndexedDB.

* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https-expected.txt: Added.
* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https.html: Added.
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js: Added.
(async createKeys):
(async addValue):
(async getValue.):
(async getValue):
(async testFetchedKeys):
(async runTest):
(async cleanupDatabase.):
(async cleanupDatabase):
(async self):
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get.js: Added.
(async test):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::wrapCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::wrapSerializedCryptoKey):
(WebCore::unwrapSerializedCryptoKey):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::wrapCryptoKey):
(WebCore::std::optional<Vector<uint8_t>>Document::unwrapCryptoKey):
(WebCore::Document::unwrapCryptoKey): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::wrapCryptoKey const): Deleted.
(WebCore::ChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::webCryptoClient):
(WebCore::Page::webCryptoClient const):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/WebCryptoClient.cpp: Added.
* Source/WebCore/page/WebCryptoClient.h: Added.
(WebCore::WebCryptoClient::wrapCryptoKey const):
(WebCore::WebCryptoClient::unwrapCryptoKey const):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::wrapCryptoKey):
(WebCore::WorkerGlobalScope::unwrapCryptoKey):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::wrapCryptoKey):
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::wrapCryptoKey):
(WebKit::WebProcessProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::_webCryptoMasterKeyForWebProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/WebCoreSupport/SWWebCryptoClient.cpp: Added.
(WebKit::SWWebCryptoClient::wrapCryptoKey const):
(WebKit::SWWebCryptoClient::unwrapCryptoKey const):
* Source/WebKit/WebProcess/WebCoreSupport/SWWebCryptoClient.h: Added.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::wrapCryptoKey const): Deleted.
(WebKit::WebChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.cpp: Added.
(WebKit::WebPageWebCryptoClient::wrapCryptoKey const):
(WebKit::WebPageWebCryptoClient::unwrapCryptoKey const):
(WebKit::WebPageWebCryptoClient::WebPageWebCryptoClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebPageWebCryptoClient.h: Added.
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
* Source/WebKitLegacy/WebCoreSupport/WebViewWebCryptoClient.h: Added.
* Source/WebKitLegacy/WebCoreSupport/WebViewWebCryptoClient.mm: Added.
(WebViewWebCryptoClient::wrapCryptoKey const):
(WebViewWebCryptoClient::unwrapCryptoKey const):
(WebViewWebCryptoClient::WebViewWebCryptoClient):
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::wrapCryptoKey const): Deleted.
(WebChromeClient::unwrapCryptoKey const): Deleted.
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
(TEST):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate webCryptoMasterKeyForWebProcess]):

Canonical link: https://commits.webkit.org/276088@main
